### DISCUSSION TITLE
How to use Custom Hooks?

### DISCUSSION THREAD

**Discussion by marcus-at-localhost** [atrocore/atrocore#397] [Q&A] [answered]
How can I setup a custom hook (e.g. calling an external web service to flush a cache or something) for entities.

I tried something like this:

```php
<?php
namespace Espo\Custom\Hooks\Product;

use Espo\ORM\Entity;

class MyHook
{
    public function afterSave(Entity $entity, array $options): void
    {
        $GLOBALS['log']->debug('Call Hook: ' . $entity->get('id'));
    }
}
```

as described here <https://docs.espocrm.com/development/hooks/>

Thanks!

**Comment by rratsun** [comment, marked as answer]
Hello,

Instead of espo hooks you should use or repository methods such as [here](https://github.com/atrocore/atropim/blob/master/app/Repositories/Attribute.php#L116), or listeners such as [here](https://github.com/atrocore/atropim/tree/master/app/Listeners). Listeners can be used only in your custom modules.

I strongly recommend you to develop your custom module and do all your modification there. How to develop you custom module you can find [here](https://github.com/atrocore/docs/blob/master/atrocore/developer-guide/creating-own-module.md).

**Reply by marcus-at-localhost** [reply]
Hi, I went through the guide and tried to wrap my head around it. But beside the slightly elaborated setup, I'm not sure how the example module, that creates a new Test Entity (or however I call it), would be able to "listen" what's going somewhere else in the system (an attribute changed etc)? Or is the listener part of the cronjob, checking constantly for changes on other entities?
In that case, I assume that the new entity needs to record every check, to know when something happened since the last check?
Am I seeing that right?
I guess in that case, I have to put it on the back burner for now (again) :D.

**Reply by MKmedicalvalues** [reply]
Hello, I am also working on a usecase, where we would like to access events from all entities, not only the ones our custom module would provide.

Is there something like the espo common hooks that we should use or is this not possible at all at the moment ?

**Reply by jnugh** [reply]
We were also not able to subscribe using a Listener class as this seems to be scoped to entities defined in the custom module itself. 
To make this work we used the `eventManager` directly from our module initialization to register our listener. I don't really understand if there is a better way to register it by convention somehow.

```php
public function onLoad()
    {
        parent::onLoad();
        $eventDispatcher = $this->container->get('eventManager');
        $eventDispatcher->getEventDispatcher()->addSubscriber(new MySubscriber());
    }
```

`MySubscriber` has to implement `EventSubscriberInterface`. In there you can subscribe every event e.g.:

```php
class YoloListener implements EventSubscriberInterface
{
    public static function getSubscribedEvents()
    {
        return ["Entity.beforeSave" => 'methodName'];
    }

    public function methodName($event)
    {
        // Do something on Entity.beforeSave
    }
}
```

**Reply by rratsun** [reply]
You have two options to create listeners. First, you can create it manually. [jnugh](https://github.com/jnugh) wrote an example. Second, is to create a special class in Listeners dir. Please, look here https://github.com/atrocore/import-feeds/tree/master/app/Listeners an examples. Import feed module have a few listeners. AttachmentController listen events triggered in Attachment Controller class. AttributeEntity listener listen an events triggered in Attribute entity. Entity listener listen events triggered in any entities. I think you should understand the logic. Another examples here https://github.com/atrocore/atropim/tree/master/app/Listeners. You can found ProductService listener, this listen an events in Product service.

Hope this help.
